### PR TITLE
feat(schema): add pyarrow Extension type to represent PortexType

### PR DESCRIPTION
**Pull request by Lee-000**

### CODE REVIEW

**Comment by GravitiRob** [comment]
Task linked: [CU-2hd6ry6 定义表示 Portex 类型的 PyArrow ExtensionType](https://app.clickup.com/t/2hd6ry6)

File: graviti/schema/pyarrow.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    Arguments:
+        portex_type: The PortexType name.
+        repo: The repo of the PortexType, including the revision.
+        storage_type: The corresponding PyArrow type.
+        kwargs: Other keyword arguments to init the PortexType.
+
+    """
+
+    def __init__(
+        self, portex_type: str, storage_type: Any, repo: Optional[str] = None, **kwargs: Any
```
Comment (linjiX) [inline graviti/schema/pyarrow.py:25]: 1. `portex_type` is a `str`? Should it be `portex_type_name`?
2. Shall split `repo` into `url` and `revision`?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/schema/pyarrow.py (hunk 1 above)
Comment (linjiX) [reply graviti/schema/pyarrow.py:25]: 1. Why `repo` is `Optional`?
2. shall we add `nullable` parameter?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/schema/pyarrow.py (hunk 1 above)
Comment (linjiX) [reply graviti/schema/pyarrow.py:25]: Why `storage_type` is `Any`?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/schema/pyarrow.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        content = {"portex_type": self._portex_type}
+        if self._repo:
+            content["repo"] = self._repo
+        if self._kwargs:
+            content.update(self._kwargs)
+        return json.dumps(content).encode("utf-8")
+
+    @classmethod
+    def __arrow_ext_deserialize__(cls, storage_type: Any, serialized: bytes) -> "PyArrowType":
+        args = json.loads(serialized.decode("utf-8"))
```
Comment (linjiX) [inline graviti/schema/pyarrow.py:42]: kwargs?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/schema/pyarrow.py (hunk 1 above)
Comment (Lee-000) [reply graviti/schema/pyarrow.py:25]: Numeric builtin types do not have repo.

Comment (linjiX) [reply graviti/schema/pyarrow.py:25]: shall we another class for builtin type?

File: graviti/schema/pyarrow.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+#
+# Copyright 2022 Graviti. Licensed under MIT License.
+#
+"""The PyArrow extension types to represent PortexType."""
+
+import json
+from typing import Any, Dict, Optional, Type, TypeVar
+
+import pyarrow as pa
+from pyarrow.lib import DataType  # pylint: disable=no-name-in-module
```
Comment (linjiX) [inline graviti/schema/pyarrow.py:11]: will it be better to put `DataType` in `TYPE_CHECKING` block？

File: graviti/schema/pyarrow.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+# https://pylint.pycqa.org/en/latest/technical_reference/c_extensions.html
+
+
+class BuiltinExtension(  # pylint: disable=too-few-public-methods
+    pa.ExtensionType  # type: ignore[misc]
```
Comment (linjiX) [inline graviti/schema/pyarrow.py:17]: why we need to ignore misc?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/schema/pyarrow.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        return cls(**kwargs, storage_type=storage_type)
+
+    def _dumps(self) -> Dict[str, Any]:
+        content = {"portex_type_name": self._portex_type_name, "nullable": self._nullable}
+        if self._kwargs:
+            content.update(self._kwargs)
+        return content
+
+
+class ExternalExtension(BuiltinExtension):  # pylint: disable=too-few-public-methods
```
Comment (linjiX) [inline graviti/schema/pyarrow.py:59]: External is a subclass of Builtin? Will it be better if External and Builtin are subclasses of a base class?

File: graviti/schema/pyarrow.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+    """
+
+    _T = TypeVar("_T", bound="BuiltinExtension")
+
+    def __init__(
+        self,
+        portex_type_name: str,
+        storage_type: DataType,
+        *,
+        nullable: bool = False,
```
Comment (linjiX) [inline graviti/schema/pyarrow.py:36]: pyarrow have `nullable` parameter? can we reuse it?

File: graviti/schema/pyarrow.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        self._portex_type_name = portex_type_name
+        self._nullable = nullable
+        self._kwargs = kwargs
+        pa.ExtensionType.__init__(self, storage_type, f"builtin.{portex_type_name}")
+
+    def __arrow_ext_serialize__(self) -> bytes:
+        return json.dumps(self._dumps()).encode("utf-8")
+
+    @classmethod
+    def __arrow_ext_deserialize__(cls: Type[_T], storage_type: Any, serialized: bytes) -> _T:
```
Comment (linjiX) [inline graviti/schema/pyarrow.py:48]: `storage_type` is `Any`?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/schema/pyarrow.py (hunk 4 above)
Comment (Lee-000) [reply graviti/schema/pyarrow.py:17]: Because `ExtensionType` has type `Any`.

File: graviti/schema/pyarrow.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    """
+
+    def __init__(
+        self,
+        portex_type_name: str,
+        storage_type: DataType,
+        *,
+        nullable: bool = False,
+        url: Optional[str] = None,
+        revision: Optional[str] = None,
```
Comment (linjiX) [inline graviti/schema/pyarrow.py:79]: why `url` and `revision` are `Optional`?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/schema/pyarrow.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+        **kwargs: Any,
+    ):
+        super().__init__(portex_type_name, storage_type, nullable=nullable, **kwargs)
+        self._url = url
+        self._revision = revision
+        pa.ExtensionType.__init__(self, storage_type, f"external.{portex_type_name}")
+
+    def _dumps(self) -> Dict[str, Any]:
```
Comment (linjiX) [inline graviti/schema/pyarrow.py:87]: the naming style in `graviti` SDK is `to_xxx`
try not to use `dump` like `tensorbay` SDK.

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/schema/pyarrow.py (hunk 6 above)
Comment (Lee-000) [reply graviti/schema/pyarrow.py:36]: Only `pyarrow.field` have the `nullable` parameter.

File: graviti/schema/pyarrow.py [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+
+    """
+
+    _T = TypeVar("_T", bound="_ExtensionBase")
+
+    _PYARROW_PACKAGE_NAME: str
+
+    def __init__(
+        self,
+        portex_type_name: str,
```
Comment (linjiX) [inline graviti/schema/pyarrow.py:36]: will it be better if `portex_type_name` -> `name`?

File: graviti/schema/pyarrow.py [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+        portex_type_name: The PortexType name.
+        storage_type: The corresponding PyArrow type.
+        nullable: Whether the PortexType is nullable.
+        kwargs: Other keyword arguments to init the PortexType.
+
+    """
+
+    _T = TypeVar("_T", bound="_ExtensionBase")
+
+    _PYARROW_PACKAGE_NAME: str
```
Comment (linjiX) [inline graviti/schema/pyarrow.py:32]: we can use `typing.ClassVar` to mark the class variable

Comment (linjiX) [reply graviti/schema/pyarrow.py:32]: Why PYARROW package? Should it be PORTEX package?

**Review by linjiX — CHANGES_REQUESTED** [review]

File: graviti/schema/pyarrow.py (hunk 11 above)
Comment (Lee-000) [reply graviti/schema/pyarrow.py:32]: Because it's the extension name in pyarrow?

**Review by linjiX — APPROVED** [review]

**Pull request merged**